### PR TITLE
[WB-7574] Improve xgboost wandb_callback

### PR DESCRIPTION
**Pull request by ayulockin**
https://wandb.atlassian.net/browse/WB-7574

Description
-----------

What does the PR do?

This PR improves and extends the `wandb_callback` for XGBoost. Here's the list of features that the new `WandbCallback` has:
    
- log the booster model configuration to Weights & Biases
- log evaluation metrics collected by XGBoost, such as rmse, accuracy, etc to Weights & Biases
- log training metric collected by XGBoost (if you provide training data to eval_set)
- log the best score and the best iteration
- save and upload your trained model to Weights & Biases Artifacts (when `log_model = True`)
- log feature importance plot when `log_feature_importance=True` (default).
- Capture the best eval metric in `wandb.summary` when `define_metric=True` (default).

Here's an example dashboard: https://wandb.ai/ayut/credit_scorecard

Testing
-------

How was this PR tested?

`yea` test.

Release Notes
-------------

Below, please enter user-facing release notes as one or more bullet points.
If your change is not user-visible, write `NO RELEASE NOTES` instead, with no bullet points.

------------- BEGIN RELEASE NOTES ------------------

NO RELEASE NOTES

------------- END RELEASE NOTES --------------------

### CODE REVIEW

**Review by morganmcg1 — APPROVED** [review]
have already reviewed and given feedback in personal fork which as been implemented, lgtm

**Review by vanpelt — COMMENTED** [review]
Looks great.  It would be really good if we could get python type annotations added, not a hard requirement but we're trying to annotate as much code as possible and it will make the user experience better in IDE's.

File: wandb/integration/xgboost/xgboost.py [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+        self.log_feature_importance = log_feature_importance
+        self.importance_type = importance_type
+        self.define_metric = define_metric
+
+    def before_training(self, model):
+        """Run before training is finished"""
+
+        # Update W&B config
+        config = model.save_config()
+        wandb.config.update(json.loads(config))
```
Comment (raubitsj) [inline wandb/integration/xgboost/xgboost.py:108]: wandb.config.update() pattern is often incompatible with sweeps and should usually be avoided.   Im not sure if that is true in this case.

File: wandb/integration/xgboost/xgboost.py [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+
+        return model
+
+    def after_training(self, model):
+        """Run after training is finished."""
+
+        # Log the booster model as artifacts
+        if self.log_model:
+            self._log_model_as_artifact(model)
```
Comment (raubitsj) [inline wandb/integration/xgboost/xgboost.py:117]: can you use log_model in your yea test

File: wandb/integration/xgboost/xgboost.py [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+
+        wandb.log({"epoch": epoch})
+
+        self.define_metric = False
+
+        return False
+
+    def _log_model_as_artifact(self, model):
+        model_name = f"{wandb.run.id}_model.json"
+        model_path = f"{wandb.run.dir}/{model_name}"
```
Comment (raubitsj) [inline wandb/integration/xgboost/xgboost.py:154]: windows?

File: wandb/integration/xgboost/xgboost.py [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+            {
+                "Feature Importance": wandb.plot.bar(
+                    table, "Feature", "Importance", title="Feature Importance"
+                )
+            }
+        )
+
+    def _define_metric(self, data, metric_name):
+        if "loss" in str.lower(metric_name):
+            wandb.define_metric(f"{data}-{metric_name}", summary="min")
```
Comment (raubitsj) [inline wandb/integration/xgboost/xgboost.py:175]: can you have "loss" in your sample yea test

File: wandb/integration/xgboost/xgboost.py [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+    warnings.warn(
+        "wandb_callback will be deprecated in favor of WandbCallback. Please use WandbCallback for more features.",
+        UserWarning,
+        stacklevel=2,
+    )
+
+    def callback(env):
+        for k, v in env.evaluation_result_list:
+            wandb.log({k: v}, commit=False)
+        wandb.log({})
```
Comment (raubitsj) [inline wandb/integration/xgboost/xgboost.py:42]: can you add an old style test in yea also.

File: wandb/integration/xgboost/xgboost.py [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+            }
+        )
+
+    def _define_metric(self, data, metric_name):
+        if "loss" in str.lower(metric_name):
+            wandb.define_metric(f"{data}-{metric_name}", summary="min")
+        elif str.lower(metric_name) in MINIMIZE_METRICS:
+            wandb.define_metric(f"{data}-{metric_name}", summary="min")
+        elif str.lower(metric_name) in MAXIMIZE_METRICS:
+            wandb.define_metric(f"{data}-{metric_name}", summary="max")
```
Comment (raubitsj) [inline wandb/integration/xgboost/xgboost.py:179]: include a maximize metric

File: wandb/integration/xgboost/xgboost.py [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+        self.define_metric = False
+
+        return False
+
+    def _log_model_as_artifact(self, model):
+        model_name = f"{wandb.run.id}_model.json"
+        model_path = f"{wandb.run.dir}/{model_name}"
+        model.save_model(str(model_path))
+
+        model_artifact = wandb.Artifact(name=model_name, type="model")
```
Comment (raubitsj) [inline wandb/integration/xgboost/xgboost.py:157]: are we doing this anywhere else? i know there are active discussions on how we want to standardize on model logging.

**Review by raubitsj — COMMENTED** [review]
A few comments.
I will also add some telemetry changes PR ontop of this branch so we can track xgboost callback

File: wandb/integration/xgboost/xgboost.py (hunk 1 above)
Comment (ayulockin) [reply wandb/integration/xgboost/xgboost.py:108]: I have tested this callback for sweeps [here](https://wandb.ai/ayut/btc_hyperparam_search?workspace=user-ayush-thakur). It ran just fine. Here's the [code](https://www.kaggle.com/ayuraj/xgboost-hyperparameter-tuning-with-w-b-sweeps/notebook) as kaggle kernel.

Comment (ayulockin) [reply wandb/integration/xgboost/xgboost.py:108]: Here's a sweep run's [config](https://wandb.ai/ayut/btc_hyperparam_search/runs/je2znhta/overview?workspace=user-ayush-thakur) differing from another sweep run's [config](https://wandb.ai/ayut/btc_hyperparam_search/runs/m35jqj0h/overview?workspace=user-ayush-thakur).

File: wandb/integration/xgboost/xgboost.py (hunk 3 above)
Comment (ayulockin) [reply wandb/integration/xgboost/xgboost.py:154]: will fix it.

File: wandb/integration/xgboost/xgboost.py (hunk 4 above)
Comment (ayulockin) [reply wandb/integration/xgboost/xgboost.py:175]: what do you mean by this?

File: wandb/integration/xgboost/xgboost.py (hunk 5 above)
Comment (ayulockin) [reply wandb/integration/xgboost/xgboost.py:42]: Sure. it will be a separate test right?

File: wandb/integration/xgboost/xgboost.py (hunk 6 above)
Comment (ayulockin) [reply wandb/integration/xgboost/xgboost.py:179]: I didn't get this? Do you wanna have me write a test for maximize metric too?

File: wandb/integration/xgboost/xgboost.py (hunk 7 above)
Comment (ayulockin) [reply wandb/integration/xgboost/xgboost.py:157]: I am familiar with Keras WandbCallback and there the best model is saved in the run page itself and not the artifacts. I am confirming this though.

File: wandb/integration/xgboost/xgboost.py (hunk 4 above)
Comment (raubitsj) [reply wandb/integration/xgboost/xgboost.py:175]: These lines are not covered by your yea test..  if you have "loss" as a metric then we will have better assurance that these lines are working as intentended.

File: wandb/integration/xgboost/xgboost.py (hunk 5 above)
Comment (raubitsj) [reply wandb/integration/xgboost/xgboost.py:42]: yes.  just a simple one, probably one in our examples docs somewhere.

File: wandb/integration/xgboost/xgboost.py (hunk 6 above)
Comment (raubitsj) [reply wandb/integration/xgboost/xgboost.py:179]: Doesnt have to be a separate test,  cant you add a metric that is named something that matches the maximize check to the existing test...    We dont have full coverage requirements but it seems you did all the work, might as well just add a few more lines to a test and get all this code covered.

File: wandb/integration/xgboost/xgboost.py (hunk 7 above)
Comment (ayulockin) [reply wandb/integration/xgboost/xgboost.py:157]: Not sure if we are doing it anywhere. (asked AC and Morgan)

File: wandb/integration/xgboost/xgboost.py (hunk 6 above)
Comment (ayulockin) [reply wandb/integration/xgboost/xgboost.py:179]: Added a metric but it's not meaningful from an ML perspective but I think that's not of a concern here. Thanks for the clarification.

File: wandb/integration/xgboost/xgboost.py (hunk 2 above)
Comment (ayulockin) [reply wandb/integration/xgboost/xgboost.py:117]: I can use `log_model` in my yea test but should I test it using yea logic?

File: wandb/integration/xgboost/xgboost.py (hunk 5 above)
Comment (ayulockin) [reply wandb/integration/xgboost/xgboost.py:42]: I have added a simple one as `yea` test itself. If that's not okay let me know.

**Comment by ayulockin** [comment]
@raubitsj I have made the changes according to your suggestions. Please let me know if it works.
